### PR TITLE
handling the partial/rawblock in non data state.

### PR DESCRIPTION
**Pull request by neraliu**
- using enum type for the AST node type
- adding the logic to warn developers if partial, rawblock, and raw expression {{&expression}} are in non-data state.

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: src/handlebars-utils.js [hunk 1]
```diff
@@ -73,6 +73,10 @@ HandlebarsUtils.SINGLE_ESCAPED_MUSTACHE = 9;
 HandlebarsUtils.DOUBLE_ESCAPED_MUSTACHE = 10;
 HandlebarsUtils.NOT_ESCAPED_MUSTACHE = 11;
 
+/* the AST node type */
+HandlebarsUtils.AST_NODE = 'node';
+HandlebarsUtils.AST_HTML = 'html';
```
Comment (adon-at-work) [inline src/handlebars-utils.js]: would you mind changing them as numbers too?

**Comment by adon-at-work** [comment]
pls git rebase this to remove dist/* files

File: src/context-parser-handlebars.js [hunk 2]
```diff
@@ -415,31 +406,46 @@ ContextParserHandlebars.prototype.analyzeAst = function(ast, contextParser, char
...
+                node.type === handlebarsUtils.COMMENT_EXPRESSION_LONG_FORM) {
+
+                // if the 'rawblock', 'partial expression' and 'ampersand expression' are not in Data State, 
+                // we should warn the developers or throw exception in strict mode
+                if ((node.type === handlebarsUtils.RAW_BLOCK ||
+                    node.type === handlebarsUtils.PARTIAL_EXPRESSION ||
+                    node.type === handlebarsUtils.AMPERSAND_EXPRESSION) && parser.getCurrentState() !== stateMachine.State.STATE_DATA) {
+                    msg = (this._config._strictMode? '[ERROR]' : '[WARNING]') + " SecureHandlebars: " + (node.type === handlebarsUtils.RAW_BLOCK? 'Raw block' : 
+                                                                                                         node.type === handlebarsUtils.PARTIAL_EXPRESSION? 'Partial expression':
+                                                                                                         node.type === handlebarsUtils.AMPERSAND_EXPRESSION? 'Raw expression': '') + ' is in non-DATA State!';
```
Comment (adon-at-work) [inline src/context-parser-handlebars.js]: can you improve this by outputting `{{>whatever}}` instead?

File: src/context-parser-handlebars.js [hunk 3]
```diff
@@ -415,31 +406,46 @@ ContextParserHandlebars.prototype.analyzeAst = function(ast, contextParser, char
...
+                node.type === handlebarsUtils.BRANCH_END_EXPRESSION ||
+                node.type === handlebarsUtils.ELSE_EXPRESSION ||
+                node.type === handlebarsUtils.COMMENT_EXPRESSION_SHORT_FORM ||
+                node.type === handlebarsUtils.COMMENT_EXPRESSION_LONG_FORM) {
+
+                // if the 'rawblock', 'partial expression' and 'ampersand expression' are not in Data State, 
+                // we should warn the developers or throw exception in strict mode
+                if ((node.type === handlebarsUtils.RAW_BLOCK ||
+                    node.type === handlebarsUtils.PARTIAL_EXPRESSION ||
+                    node.type === handlebarsUtils.AMPERSAND_EXPRESSION) && parser.getCurrentState() !== stateMachine.State.STATE_DATA) {
```
Comment (adon-at-work) [inline src/context-parser-handlebars.js]: can you separate these 3 ifs as another `else if` in one layer (line 429) up

File: src/handlebars-utils.js (hunk 1 above)
Comment (adon-at-work) [inline src/handlebars-utils.js]: kudos for the clean up :)

**Comment by adon-at-work** [comment]
excellent! :+1:

**Pull request merged**